### ISSUE TITLE
Cannot get UID/GID

### ISSUE THREAD

**Issue by robbyemmert** [npm/uid-number#7]
I'm getting the following error using node-alpine:10.9.0.

It seems like lots of other people are also having similar issues on a variety of different node versions, but there is currently not an open issue (that I am aware of) representing this issue with recent examples.

For reference, please also see: https://github.com/npm/uid-number/issues/3

Here is the error I'm getting:
```
Error: could not get uid/gid
[ 'nobody', 0 ]

    at /usr/local/lib/node_modules/npm/node_modules/uid-number/uid-number.js:37:16
    at ChildProcess.exithandler (child_process.js:296:5)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:961:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:250:5)
TypeError: Cannot read property 'get' of undefined
    at errorHandler (/usr/local/lib/node_modules/npm/lib/utils/error-handler.js:205:18)
    at /usr/local/lib/node_modules/npm/bin/npm-cli.js:76:20
    at cb (/usr/local/lib/node_modules/npm/lib/npm.js:228:22)
    at /usr/local/lib/node_modules/npm/lib/npm.js:266:24
    at /usr/local/lib/node_modules/npm/lib/config/core.js:83:7
    at Array.forEach (<anonymous>)
    at /usr/local/lib/node_modules/npm/lib/config/core.js:82:13
    at f (/usr/local/lib/node_modules/npm/node_modules/once/once.js:25:25)
    at afterExtras (/usr/local/lib/node_modules/npm/lib/config/core.js:173:20)
    at Conf.<anonymous> (/usr/local/lib/node_modules/npm/lib/config/core.js:231:22)
/usr/local/lib/node_modules/npm/lib/utils/error-handler.js:205
  if (npm.config.get('json')) {
                 ^

TypeError: Cannot read property 'get' of undefined
    at process.errorHandler (/usr/local/lib/node_modules/npm/lib/utils/error-handler.js:205:18)
    at process.emit (events.js:182:13)
    at process._fatalException (internal/bootstrap/node.js:472:27)
```

Here is my docker file for reference:
```
FROM node:10.9.0-alpine

WORKDIR /usr/src/app

# Install app dependencies
# A wildcard is used to ensure both package.json AND package-lock.json are copied
# where available (npm@5+)
COPY package*.json ./

# Install a bunch of junk I need
RUN apk --no-cache add --virtual native-deps g++ gcc libgcc libstdc++ linux-headers make python
# This is the line that fails
RUN npm install --quiet node-gyp -g
# I never make it to here
RUN npm install
RUN apk del native-deps

# If you are building your code for production
# RUN npm install --only=production

# Bundle app source
COPY . .

# Setup the app
RUN npm run build

EXPOSE 1337

CMD [ "npm", "start" ]
```

**Comment by superdevofficial** [comment]
Hi !

A workaround, you can add this line to your dockerfile, before invoke npm :
`RUN npm config set unsafe-perm true`

More details here :
https://github.com/npm/uid-number/issues/3

**Comment by knoxcard** [comment]
Which console are you running this code under: Windows, Windows Bash, or Linux?

**Comment by vovimayhem** [comment]
@knoxcard That would be Linux plain ol' SH (`!# /bin/sh`)

We're using Alpine Linux :)

**Comment by keblodev** [comment]
the issue is still happening.
And apparently it's somehow related to types of instances or their sizes, or location...
running from `node:8.4.0-alpine`

**same exact** docker file runs just fine on staging cluster in Taiwan.
but is failing with this error on prod cluster in another asian region.

wish there was at least **any** explanation to the issue from maintainers

**Comment by FewKinG** [comment]
npm depending on uid-number requiring using unsafe mode as a workaround for 4+ years, maintainers don't care to even explain. Don't wanna blame anyone (should probably try to fix it myself), but for an open source eco-system this is not a good sign.

**Comment by ajutzeler** [comment]
I am experiencing the same problem:
Depending on the docker host it works or it fails.

**Comment by phillycheeze** [comment]
See here for more reference on the issue if you're using the M5 instance types in AWS: https://bitbucket.org/site/master/issues/16846/npm-error-eacces-permission-denied

**Comment by isaacs** [comment]
This will be addressed in the next release (6.11.0) with better error handling and a more useful error message.

npm version 7 will abandon the "run scripts as `nobody`" approach entirely.  It doesn't provide much security against realistic threat models, and just makes a lot of difficulty for folks.

From the 6.11.0 changelog:

* [`8b43c9624`](https://github.com/npm/cli/commit/8b43c962498c8e2707527e4fca442d7a4fa51595)
  [nodejs/node#28987](https://github.com/nodejs/node/issues/28987)
  [npm.community#6032](https://npm.community/t/npm-ci-doesnt-respect-npmrc-variables/6032)
  [npm.community#6658](https://npm.community/t/npm-ci-doesnt-fill-anymore-the-process-env-npm-config-cache-variable-on-post-install-scripts/6658)
  [npm.community#6069](https://npm.community/t/npm-ci-does-not-compile-native-dependencies-according-to-npmrc-configuration/6069)
  [npm.community#9323](https://npm.community/t/npm-6-9-x-not-passing-environment-to-node-gyp-regression-from-6-4-x/9323/2)
  Fix the regression where random config values in a .npmrc file are not
  passed to lifecycle scripts, breaking build processes which rely on them.
  ([@isaacs](https://github.com/isaacs))

**Comment by isaacs** [comment]
Er, whoops!  Wrong changelog entry.  I mean this one:

* [`769d2e057`](https://github.com/npm/cli/commit/769d2e057daf5a2cbfe0ce86f02550e59825a691)
  Better error on invalid --user/--group This addresses the issue when
  people fail to install binary packages on Docker and other environments
  where there is no 'nobody' user.  ([@isaacs](https://github.com/isaacs))
